resources/js/beatmap-discussions/main.tsx

```diff
@@ -0,0 +1,253 @@
+import * as React from 'react';
+import BeatmapsetDiscussionsShowStore from 'stores/beatmapset-discussions-show-store';
+import { defaultFilter, parseUrl, stateFromDiscussion } from 'utils/beatmapset-discussion-helper';
+import { parseJson, storeJson } from 'utils/json';
+import { nextVal } from 'utils/seq';
+import { currentUrl } from 'utils/turbolinks';
+import { Discussions } from './discussions';
+import DiscussionsState from './discussions-state';
+import { Events } from './events';
+import { Header } from './header';
+import { ModeSwitcher } from './mode-switcher';
+import { NewDiscussion } from './new-discussion';
+
+const checkNewTimeoutDefault = 10000;
+const checkNewTimeoutMax = 60000;
+const beatmapsetJsonId = 'json-beatmapset';
+
+interface Props {
+  reviewsConfig: {
+    max_blocks: number;
+  };
+}
+
+interface UpdateResponseJson {
+  beatmapset: BeatmapsetWithDiscussionsJson;
+}
+
+@observer
+export default class Main extends React.Component<Props> {
+  @observable private readonly discussionsState: DiscussionsState;
+  private readonly disposers = new Set<((() => void) | undefined)>();
+  private readonly eventId = `beatmap-discussions-${nextVal()}`;
+  // FIXME: update url handler to recognize this instead
+  private readonly focusNewDiscussion = currentUrl().hash === '#new';
+  private readonly modeSwitcherRef = React.createRef<HTMLDivElement>();
+  private readonly newDiscussionRef = React.createRef<HTMLDivElement>();
+  private nextTimeout = checkNewTimeoutDefault;
+  @observable private readonly store;
+  private timeoutCheckNew?: number;
+  private xhrCheckNew?: JQuery.jqXHR<UpdateResponseJson>;
+
+  constructor(props: Props) {
+    super(props);
+
+    // TODO: avoid reparsing/loading everything on browser navigation for better performance.
+    const beatmapset = parseJson<BeatmapsetWithDiscussionsJson>(beatmapsetJsonId);
+
+    this.store = new BeatmapsetDiscussionsShowStore(beatmapset);
+    this.discussionsState = new DiscussionsState(this.store);
+
+    makeObservable(this);
+  }
+
+  componentDidMount() {
+    $(document).on(`ajax:success.${this.eventId}`, '.js-beatmapset-discussion-update', this.ujsDiscussionUpdate);
+    $(document).on(`click.${this.eventId}`, '.js-beatmap-discussion--jump', this.jumpToClick);
+    $(document).on(`turbolinks:before-cache.${this.eventId}`, this.saveState);
+
+    if (this.discussionsState.jumpToDiscussion) {
+      this.disposers.add(core.reactTurbolinks.runAfterPageLoad(this.jumpToDiscussionByHash));
+    }
+
+    this.timeoutCheckNew = window.setTimeout(this.checkNew, checkNewTimeoutDefault);
+  }
+
+  componentWillUnmount() {
+    $.unsubscribe(`.${this.eventId}`);
+    $(document).off(`.${this.eventId}`);
+
+    document.documentElement.style.removeProperty('--scroll-padding-top-extra');
+
+    window.clearTimeout(this.timeoutCheckNew);
+    this.xhrCheckNew?.abort();
+    this.disposers.forEach((disposer) => disposer?.());
+    this.discussionsState.destroy();
```

Comment:
I think these should also be run on `turbolinks:before-cache`. the `destroy()` notably involves removing url handler which means if the state somehow changes between new page load and unmount, the url will be wrong